src/lib.rs

```diff
@@ -0,0 +1,257 @@
+                        Ok(p) => {
+                            *this.param = Some(p);
+                            *this.state = State::Closed;
+                            return Poll::Ready(Ok(()))
+                        }
+                        Err(e) => return Poll::Ready(Err(e))
+                    }
+                State::Closed => return Poll::Ready(Ok(()))
+            }
+        }
+    }
+
+    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<(), Self::Error>> {
+        loop {
+            let mut this = self.as_mut().project();
+            match this.state {
+                State::Empty =>
+                    if let Some(p) = this.param.take() {
+                        let future = (this.lambda)(p, Action::Close);
+                        this.future.set(Some(future));
+                        *this.state = State::Closing;
+                    } else {
+                        return Poll::Ready(Ok(()))
+                    }
+                State::Sending =>
+                    match ready!(this.future.as_pin_mut().unwrap().poll(cx)) {
+                        Ok(p) => {
+                            *this.param = Some(p);
+                            *this.state = State::Empty
+                        }
+                        Err(e) => return Poll::Ready(Err(e))
+                    }
+                State::Flushing =>
+                    match ready!(this.future.as_pin_mut().unwrap().poll(cx)) {
+                        Ok(p) => {
+                            *this.param = Some(p);
+                            *this.state = State::Empty
+                        }
+                        Err(e) => return Poll::Ready(Err(e))
+                    }
+                State::Closing =>
+                    match ready!(this.future.as_pin_mut().unwrap().poll(cx)) {
+                        Ok(p) => {
+                            *this.param = Some(p);
+                            *this.state = State::Closed;
+                            return Poll::Ready(Ok(()))
+                        }
+                        Err(e) => return Poll::Ready(Err(e))
+                    }
+                State::Closed => return Poll::Ready(Ok(()))
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use async_std::{io, task};
+    use futures::{prelude::*, stream};
+    use crate::{Action, make_sink};
+
+    #[test]
+    fn smoke_test() {
+        let sink = make_sink(io::stdout(), |mut stdout, action| async move {
+            match action {
+                Action::Send(x) => stdout.write_all(x).await?,
+                Action::Flush => stdout.flush().await?,
+                Action::Close => stdout.close().await?
+            }
+            Ok::<_, io::Error>(stdout)
+        });
+
+        task::block_on(async move {
+            let values = vec![Ok(&b"hello\n"[..]), Ok(&b"world\n"[..])];
+            assert!(stream::iter(values).forward(sink).await.is_ok())
```

Comment:
Wouldn't it make more sense to use something other than `io::stdout` to check whether `["hello", "world"]` comes out on the other end?

Reply:
Absolutely.